modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java

```diff
@@ -39,13 +39,117 @@
 import java.io.IOException;
 import java.text.SimpleDateFormat;
 import java.util.*;
+import java.util.function.Function;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import static org.openapitools.codegen.utils.StringUtils.camelize;
 import static org.openapitools.codegen.utils.StringUtils.underscore;
 
 public abstract class AbstractTypeScriptClientCodegen extends DefaultCodegen implements CodegenConfig {
+
+    @SuppressWarnings("StringBufferField")
+    public static class ParameterExpander {
+        public enum STRATEGY {
+            legacy("Old legacy implementation that only supports the 'simple' style"
+                           + " and only calls encodeURIComponent()"
+                           + " (as opposed to encoding of all restricted characters of RFC 6570)"),
+            custom("Allows for providing a custom parameter encoding function via customization of " +
+                           "the 'api.service.mustache' template.");
```

Comment:
I suggest to create a separate .mustache template, so that a consumer does not need to copy the entire api.service.mustache in order to customize that.

As an alternative, can we use the `Configuration` object to pass suche a custom parameter encoding function?

Reply:
> can we use the Configuration object to pass such a custom parameter encoding function?

veeeery nice idea, didn't think of that :facepalm: 

Lets say, I'd create the Configuration property `paramEncoder?: (param: Parameter) => string;`

There are multiple ways of implementing this: 
1. *keep* the new Cli-Param `paramExpansionStrategy` and only generate the encoder-stuff if the cli-param is set.
2. *drop* the new Cli-Param and evaluate the Configuration property mentioned above
  a) implement this in a way that api.service.mustache is still unchanged (this will make the java code uglier)
  b) api.service.mustache may be changed so that the new behavior is implemented nicely

Option 1 might be implemented in a way so current generator setups may run unmodified. As such, this might be released in the next minor version.
Option 2 is (imho) a breaking change as customized mustache templates need to be rebased to the new versions (at the very least configuration.mustache and probably api.service.mustache)

Depending how much you value 100% backwards compatibility vs future usability vs maintainability, all 3 options provide benefits and drawbacks.

Please choose :)

P.S.: I'd be happy to implement multiple versions if you want to try different approaches!

Reply:
thanks for evaluating the options

I would go with option 2b, since overriding templates is not guaranteed to be backwards compatible